src/Auth0.OidcClient.Core/Tokens/AsymmetricSignatureVerifier.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+
+namespace Auth0.OidcClient.Tokens
+{
+    internal class AsymmetricSignatureVerifier : ISignatureVerifier
+    {
+        private readonly IList<JsonWebKey> keys;
+
+        public AsymmetricSignatureVerifier(IList<JsonWebKey> keys)
+        {
+            this.keys = keys;
+        }
+
+        public JwtSecurityToken VerifySignature(string token)
+        {
+            var securityTokenHandler = new JwtSecurityTokenHandler();
+
+            JwtSecurityToken decoded;
+            try
+            {
+                decoded = securityTokenHandler.ReadJwtToken(token);
+            }
+            catch (ArgumentException e)
+            {
+                throw new IdTokenValidationException("ID token could not be decoded.", e);
+            }
+
+            if (decoded.Header.Alg != "RS256")
+                throw new IdTokenValidationException($"Signature algorithm of \"{decoded.Header.Alg }\" is not supported. Expected the ID token to be signed with \"RS256\".");
+
+            var publicKey = FindPublicKeyByKid(decoded.Header.Kid);
```

Comment:
Would this throw an error if the token did not have a `kid` in the header or would it just compare to `null` or similar?

Reply:
Is that a likely scenario?  Should probably have a cross-SDK plan for it and ensure it's consistently applied. Right now it just means it will always fail with an error saying it couldn't find the kid of null.

Reply:
It's fine to fail when you couldn't find a public key related to that key id you are looking for, even if that value is null to begin with. Auth0 use cases at least will **always** send the `kid` claim on the header. Rest assured :)

Reply:
Will leave as-is then.